README

Shadow

A zero-configuration YAML RESTful ActiveRecord server.

== License

Copyright (c) 2007 Cloudburst, LLC. See the included LICENSE 
file.

== Installation

  sudo gem install shadow

== Usage

Shadow requires a Rails-style <tt>database.yml</tt> configuration.

Start the server with:

  shadow [path_to/database.yml] [environment] [route]

Now you can query it at:

  http://0.0.0.0:2001/route/table_name/record_id
  
You need to make RESTful HTTP requests (GET, POST, PUT, or DELETE), with YAML form payloads for POST and PUT.

== Further resources

* http://blog.evanweaver.com/pages/code#shadow
* http://rubyforge.org/forum/forum.php?forum_id=15381


CHANGELOG

v1.1. Documentation.

v1. Everything except documentation.


test/schema.rb

module ActiveRecord

  Schema.define(:version => 0) do
    create_table :cats, :force => true do |t|
      t.column :name, :string
      t.column :size, :integer
      t.column :created_at, :datetime, :null => false
      t.column :updated_at, :datetime, :null => false
    end
  
    create_table :dogs, :force => true do |t|
      t.column :name, :string
      t.column :size, :integer
      t.column :created_at, :datetime, :null => false
      t.column :updated_at, :datetime, :null => false
    end
  end

  class Cat < Base; end
  Cat.create(:name => "Blue", :size => 3)
  Cat.create(:name => "Tom", :size => 6)
  
  class Dog < Base; end
  Dog.create(:name => "Rover", :size => 5)
  Dog.create(:name => "Spot", :size => 5)

end


test/integration/test_shadow.rb
#!/usr/bin/env ruby

require 'rubygems'
require 'test/unit'
require 'rfuzz/session'

DIR = File.dirname(__FILE__)
require "#{DIR}/../../lib/shadow"

class ShadowText < Test::Unit::TestCase
  APP, HOST, PORT = "menagerie", "0.0.0.0", 2001
  $client = RFuzz::HttpClient.new(HOST, PORT)
  $server = Shadow.new("#{DIR}/../database.yml", "test", APP, HOST, PORT, 1)
  require "#{DIR}/../schema"
    
  def test_get
    # successful
    response = $client.get("/#{APP}/cats/1")
    assert_match /name: Blue/, response.http_body
    assert_equal "text/yaml", response["CONTENT_TYPE"]
    assert_equal "200", response.http_status
    # missing
    response = $client.get("/#{APP}/cats/25")
    assert_match /Couldn't find/, response.http_body
    assert_equal "400", response.http_status
  end
  
  def test_put
    response = $client.put("/#{APP}/cats", :body => {"name" => "Fluffy", "size" => 4}.to_yaml)
    assert YAML.load(response.http_body)["id"]
  end
  
  def test_post
    response = $client.put("/#{APP}/cats/2", :body => {"name" => "Tabby"}.to_yaml)
    assert_equal "Tabby", YAML.load(response.http_body)["name"]
    response = $client.get("/#{APP}/cats/2")
    assert_equal "Tabby", YAML.load(response.http_body)["name"]
  end
  
  def test_delete
    response = $client.delete("/#{APP}/dogs/1")
    assert_match /name: Rover/, response.http_body
    response = $client.delete("/#{APP}/dogs/1")
    assert_match /Couldn't find/, response.http_body
    assert_equal "400", response.http_status
  end
  
  def test_concurrent_reads
    correct_result = $client.get("/#{APP}/cats/1").http_body
    assert_nothing_raised do
      pids = []
      100.times do
        pids << fork do # can't use Thread because RFuzz isn't threadsafe?
          result = $client.get("/#{APP}/cats/1").http_body
          puts "Response error: #{result}" unless correct_result == result 
        end 
      end
      pids.each {|pid| Process.wait(pid)}
    end
  end

  def test_concurrent_writes
    assert_nothing_raised do
      pids = []
      100.times do
        pids << fork do
          result = $client.put("/#{APP}/dogs", 
                                        :body => {"name" => Process.pid.to_s}.to_yaml)
          id = YAML.load(result.http_body)["id"]
          sleep(2)
          result = $client.get("/#{APP}/dogs/#{id}").http_body
          puts "Response error: #{result.inspect}" unless result =~ /#{Process.pid}/
        end 
      end
      pids.each {|pid| Process.wait(pid)}
    end
  end
    
  def test_zzzz_im_asleep_zzzzz
    Process.kill(9, $server.pid)
    assert "dead"
  end
    
end


lib/shadow.rb

require 'rubygems'
require 'mongrel'
require 'active_record'
require 'sync'

class Shadow < Mongrel::HttpHandler
  attr_reader :pid

  def initialize(config, environment, name, address = '0.0.0.0', port = '2001', sleep = nil)
    ActiveRecord::Base.establish_connection(YAML.load_file(config)[environment.to_s])
    ActiveRecord::Base.allow_concurrency = false # AR doesn't release connections fast enough
    @sleep = sleep
    @sync = Sync.new
    @pid = serve(self, name, address, port)
  end
  
  # Implement the <tt>mongrel</tt> event handler. Responds to all four HTTP methods.  
  def process(request, response)
    sleep(rand * @sleep) if @sleep
    table, id = request.params["PATH_INFO"].split("/")

    obj, code = nil, 200
    @sync.synchronize(:EX) do # sad
      begin
        obj = find(table, id)
        case request.params["REQUEST_METHOD"]
          when "PUT", "POST"
            obj.update_attributes(YAML.load(request.body.read))
            obj.save!
          when "DELETE"
            obj.destroy
        end
        obj = obj.attributes
      rescue Object => e
        obj, code = e.to_s, 400
      end
    end
    
    response.start(code) do |head, out|
      head["Content-Type"] = "text/yaml"
      out.write obj.to_yaml
    end
  end
  
  # Finds and returns an ActiveRecord instance (returns a new record if <tt>id</tt> is nil). Dynamically instantiates an ActiveRecord parent class for each call. 
  def find(table, id)
    klass = Class.new(ActiveRecord::Base) { self.table_name = table }
    id ? klass.find(id) : klass.new
  end
    
  # Configure mongrel and start an instance of ourselves.      
  def serve(me, name, address, port)
    fork do
      Mongrel::Configurator.new :host => address, :pid_file => "/tmp/shadow.#{name}.pid" do
        listener :port => port do
          puts "** Serving at #{address}:#{port}/#{name}/ (pid #{Process.pid})"
          uri "/#{name}/", :handler => me
          setup_signals or run and write_pid_file and join
        end
      end
    end
  end

  private
  
  def self.d
    require 'ruby-debug'; Debugger.start; debugger
  end  
  
end
